Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -2315,16 +2321,21 @@ def %(unpickle_func_name)s(__pyx_type, long __pyx_checksum, __pyx_state):
 
             pickle_func = TreeFragment("""
                 def __reduce_cython__(self):
+                    cdef Py_ssize_t idx
                     cdef tuple state
                     cdef object _dict
-                    cdef bint use_setstate
-                    state = (%(members)s)
-                    _dict = getattr(self, '__dict__', None)
+                    cdef bint use_setstate = False
+                    with CRITICAL_SECTION(self):
+                        _dict = getattr(self, '__dict__', None)
+                        state = (%(members)s)
```

Comment:
I'm not sure if `getattr()` is something that is particularly well guarded by a critical section, but I guess that if it gets released for the lookup, it would get re-acquired after the access and for the subsequent member reads. I'm not sure that we can do better.